File: NodeJS/human_behavior.js
const { chromium } = require('playwright-extra');
const Stealth = require('playwright-extra-plugin-stealth');
chromium.use(Stealth());

function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

async function humanBehavior(page) {
  // random mouse moves
  for (let i = 0; i < 5; i++) {
    const x = Math.floor(Math.random() * 700) + 100;
    const y = Math.floor(Math.random() * 500) + 100;
    await page.mouse.move(x, y, { steps: Math.floor(Math.random() * 20) + 5 });
    await sleep(Math.random() * 400 + 200);
  }

  // scroll down in chunks
  for (let i = 0; i < 3; i++) {
    await page.evaluate(() => {
      window.scrollBy(0, window.innerHeight * 0.8);
    });
    await sleep(Math.random() * 700 + 500);
  }

  // pause and scroll up
  await sleep(Math.random() * 500 + 500);
  await page.evaluate(() => {
    window.scrollBy(0, -window.innerHeight * 0.5);
  });
  await sleep(Math.random() * 500 + 300);

  // random click
  const bx = Math.floor(Math.random() * 700) + 50;
  const by = Math.floor(Math.random() * 500) + 50;
  await page.mouse.click(bx, by);
  await sleep(Math.random() * 1000 + 1000);
}

(async () => {
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext();
  const page = await context.newPage();

  await page.goto('https://example.com');
  await humanBehavior(page);
  // continue with your scraping...
  await browser.close();
})();

File: Python/human_behavior.py
import asyncio
import random
from playwright.async_api import async_playwright
from playwright_stealth import stealth_async

async def human_behavior(page):
    # random mouse moves
    for _ in range(5):
        x = random.randint(100, 800)
        y = random.randint(100, 600)
        await page.mouse.move(x, y, steps=random.randint(5, 25))
        await asyncio.sleep(random.uniform(0.2, 0.6))

    # scroll down in chunks
    for _ in range(3):
        await page.evaluate("""() => {
            window.scrollBy(0, window.innerHeight * 0.8);
        }""")
        await asyncio.sleep(random.uniform(0.5, 1.2))

    # small pause, then scroll up
    await asyncio.sleep(random.uniform(0.5, 1.0))
    await page.evaluate("""() => {
        window.scrollBy(0, -window.innerHeight * 0.5);
    }""")
    await asyncio.sleep(random.uniform(0.3, 0.8))

    # click a random point
    bx = random.randint(50, 750)
    by = random.randint(50, 550)
    await page.mouse.click(bx, by)
    await asyncio.sleep(random.uniform(1.0, 2.0))


async def main():
    async with async_playwright() as pw:
        browser = await pw.chromium.launch(headless=True)
        context = await browser.new_context()
        page = await context.new_page()

        await stealth_async(page)  # optional stealth
        await page.goto("https://example.com")
        await human_behavior(page)
        # now you can scrape or do more...
        await browser.close()

if __name__ == "__main__":
    asyncio.run(main())
